crates/pallet-rewards/src/lib.rs

```diff
@@ -97,40 +188,265 @@ mod pallet {
 
-    fn do_finalize(_block_number: BlockNumberFor<T>) {
-        let reward = T::VoteReward::get();
+    fn reference_subsidy_for_block(
+        params: IssuanceComponentParams<BlockNumberFor<T>, BalanceOf<T>>,
+        block_height: BlockNumberFor<T>,
+    ) -> BalanceOf<T> {
+        if block_height < params.initial_subsidy_start_block {
+            return Zero::zero();
+        }
+
+        let decay_start_block = params
+            .initial_subsidy_start_block
+            .saturating_add(params.initial_subsidy_duration);
+        let t = match block_height.checked_sub(&decay_start_block) {
+            Some(t) => Self::block_number_to_balance(t),
+            None => {
+                return params.initial_subsidy;
+            }
+        };
+
+        let total_initial_subsidy = Self::block_number_to_balance(params.initial_subsidy_duration)
+            .saturating_mul(params.initial_subsidy);
+        let maybe_max_decay_issuance = params
+            .max_component_issuance
+            .checked_sub(&total_initial_subsidy);
+        let Some(max_decay_issuance) = maybe_max_decay_issuance else {
+            warn!(
+                "`max_decay_issuance` underflow when calculating reference subsidy, incorrect \
+                issuance parameters"
+            );
 
-        for voter in T::FindVotingRewardAddresses::find_voting_reward_addresses() {
-            let _imbalance = T::Currency::deposit_creating(&voter, reward);
-            T::OnReward::on_reward(voter.clone(), reward);
+            return Zero::zero();
+        };
 
-            Self::deposit_event(Event::VoteReward { voter, reward });
+        // Code below is the same as:
+        // ```
+        // k = params.initial_subsidy / max_decay_issuance
+        // params.initial_subsidy*(1 - k*t + (k*t)^2/2 - (k*t)^3/6 + (k*t)^4/24 - (k*t)^5/120)
+        // ```
+        // But written with just integers.
+        //
+        // Step 1 (multiply everything by `initial_subsidy`):
+        // ```
+        // is = params.initial_subsidy
+        // mdi = max_decay_issuance
+        // is
+        //   - is*is/mdi*t
+        //   + is*(is/mdi*t)^2/2
+        //   - is*(is/mdi*t)^3/6
+        //   + is*(is/mdi*t)^4/24
+        //   - is*(is/mdi*t)^5/120
+        // ```
+        // Step 2 (expand powers):
+        // ```
+        // is
+        //   - is*is/mdi*t
+        //   + is*is*is/mdi/mdi*t*t/2
+        //   - is*is*is*is/mdi/mdi/mdi*t*t*t/6
+        //   + is*is*is*is*is/mdi/mdi/mdi/mdi*t*t*t*t/24
+        //   - is*is*is*is*is*is/mdi/mdi/mdi/mdi/mdi*t*t*t*t*t/120
+        // ```
+        // Step 3 (reorder to avoid integer overflow or zeroing):
+        // ```
+        // is
+        //   - is*is*t/mdi
+        //   + is*is*t/mdi*is*t/mdi/2
+        //   - is*is*t/mdi*is*t/mdi*is*t/mdi/6
+        //   + is*is*t/mdi*is*t/mdi*is*t/mdi*is*t/mdi/24
+        //   - is*is*t/mdi*is*t/mdi*is*t/mdi*is*t/mdi*is*t/mdi/120
+        // ```
+
+        let is = params.initial_subsidy;
```

Comment:
Can we reduce computation (at cost of more memory) to save intermediate power terms?  i.e
```
let term=is*t/mdi;
let term2=term*term;
let term3=term2*term;
let term4=term3*term;
let term5=term4*term;
let maybe_subsidy = try {
            is.checked_sub(&(is *term))?
                .checked_add(&(is * term2 / 2_u32.into()))?
                .checked_sub(&(is * term3 / 6_u32.into()))?
                .checked_add(
                    &(is * term4
                        / 24_u32.into()),
                )?
                .checked_sub(
                    &(is * term5
                        / 120_u32.into()),
                )?
```
It does multiplication and division in the same order as you did to avoid overflow.